src/trace-plugin-loader.ts

```diff
@@ -16,278 +16,485 @@
+
+              // Get the module version.
+              let version = this.getVersion(baseDir);
+              if (version) {
+                // Warn for pre-releases.
+                if (!!semver.prerelease(version)) {
+                  if (isMainModule) {
+                    this.logger.warn(`PluginLoader#onRequire: [${name}@${
+                        version}] This module is in pre-release. Applying plugin anyways.`);
                   }
-                });
-                if (Object.keys(patchSet).length === 0) {
-                  logger.warn(`${moduleRoot}: version ${
-                      version} not supported by plugin.`);
+                  version = version.split('-')[0];
                 }
-              }
-            } else {
-              logger.warn(
-                  `${moduleRoot}: invalid version ${version}; not patching.`);
-            }
-            instrumentation.patches[moduleRoot] = patchSet;
-          }
 
-          for (const file of Object.keys(patchSet)) {
-            const patch = patchSet[file];
-            const loadPath =
-                moduleRoot ? path.join(moduleRoot, patch.file) : patch.file;
-            if (!patch.module) {
-              patch.module = originalModuleLoad(loadPath, module, false);
-            }
-            if (isInternalPatch(patch)) {
-              patch.patch(patch.module, instrumentation.agent);
-            }
-            if (isInternalIntercept(patch)) {
-              patch.module =
-                  patch.intercept(patch.module, instrumentation.agent);
-              intercepts[loadPath] = {interceptedValue: patch.module};
+                // Apply each supported plugin.
+                const plugin = this.pluginMap.get(name);
+                if (plugin) {
+                  if (plugin.isSupported(version)) {
+                    exportedValue =
+                        plugin.applyPlugin(exportedValue, file, version!);
+                  } else {
+                    this.logger.warn(`PluginLoader#onRequire: [${name}@${
+                        version}] This module is not supported by the configured set of plugins.`);
+                  }
+                }
+              } else if (isMainModule) {
+                this.logger.error(`PluginLoader#activate: [${
+                    name}] This module's version could not be determined. Not applying plugins.`);
+              }
             }
           }
-          const rootPatch = patchSet[''];
-          if (rootPatch && isInternalIntercept(rootPatch)) {
-            return rootPatch.module;
-          } else {
-            return null;
-          }
         }
+        return exportedValue;
+      });
+      this.internalState = PluginLoaderState.ACTIVATED;
+      this.logger.info(`PluginLoader#activate: Activated.`);
+    } else if (this.internalState === PluginLoaderState.DEACTIVATED) {
+      throw new Error('Currently cannot re-activate plugin loader.');
```

Comment:
Why is it so?

Reply:
There is currently no code path that causes it to be re-activated, because the Trace Agent can't be re-enabled after being disabled. So I didn't bother writing an implementation.